dashbuilder/dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/component/ExternalComponentDispatcher.java

```diff
@@ -0,0 +1,179 @@
+
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import elemental2.dom.DomGlobal;
+import elemental2.dom.MessageEvent;
+import jsinterop.base.Js;
+import org.dashbuilder.displayer.client.component.function.ComponentFunctionLocator;
+import org.dashbuilder.displayer.client.component.function.ExternalComponentFunction;
+import org.dashbuilder.displayer.client.resources.i18n.CommonConstants;
+import org.dashbuilder.displayer.external.ExternalComponentMessage;
+import org.dashbuilder.displayer.external.ExternalComponentMessageHelper;
+import org.dashbuilder.displayer.external.ExternalComponentMessageType;
+import org.dashbuilder.displayer.external.FunctionCallRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Listen to all component messages and find the destination.
+ *
+ */
+@ApplicationScoped
+public class ExternalComponentDispatcher {
+
+    private static final CommonConstants i18n = CommonConstants.INSTANCE;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExternalComponentDispatcher.class);
+
+    @Inject
+    ExternalComponentMessageHelper messageHelper;
+
+    @Inject
+    ComponentFunctionLocator functionLocator;
+
+    Set<ExternalComponentListener> listeners;
+
+    @PostConstruct
+    public void setup() {
+        listeners = new HashSet<>();
+        startListening();
+    }
+
+    public void register(ExternalComponentListener listener) {
+        // make the component listener ready by default
+        listener.prepare();
+        listeners.add(listener);
+    }
+
+    public void unregister(ExternalComponentListener listener) {
+        listeners.remove(listener);
+    }
+
+    public void onMessage(ExternalComponentMessage message) {
+
+        ExternalComponentMessageType type = messageHelper.messageType(message);
+
+        switch (type) {
+            case FILTER:
+                handleFilter(message);
+                break;
+
+            case FUNCTION_CALL:
+                handleFunction(message);
+                break;
+
+            case READY:
+                handleReady(message);
+                break;
+
+            case FIX_CONFIGURATION:
+                handleConfiguration(message);
+                break;
+
+            case CONFIGURATION_OK:
+                handleOkConfiguration(message);
+                break;
+
+            default:
+                break;
+        }
+
+    }
+
+    private void handleOkConfiguration(ExternalComponentMessage message) {
+        findDestination(message, ExternalComponentListener::configurationOk);
+    }
+
+    private void handleConfiguration(ExternalComponentMessage message) {
+        findDestination(message, destination -> {
+            Optional<String> configurationIssue = messageHelper.getConfigurationIssue(message);
```

Comment:
I would prefer to use the capability of optional instead of specific if else statement.
We could reduce this in a single line of code:
messageHelper.getConfigurationIssue(message)
.ifPresent(x-> destination.onConfigurationIssue(configurationIssue.get()))
orElse(x-> destination.onConfigurationIssue(i18n.componentConfigDefaultMessage());

Reply:
same implementation could apply in multiple places

Reply:
Hello Savitha,

Thanks for the feedback. I modified this part to be an one liner. However, in other parts it does not apply because when the result is not present I have to do operations with other object.

When we support Java 11 I wish I could refactor to use ifPresentOrElse!